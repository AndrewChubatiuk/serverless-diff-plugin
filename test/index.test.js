const AWS = require('aws-sdk-mock');
const fs = require('fs-extra');
const Plugin = require('../lib/index');

const slsDir = '.serverless';
const templatePrefix = `${slsDir}/cloudformation-template-update-stack`;
const exampleTemplate = `${templatePrefix}.json`;
const exampleOrgTemplate = `${templatePrefix}.org.json`;
const slsDefaults = {
  service: {
    service: 'foo',
    defaults: {
      stage: 'foo',
      region: 'eu-west-1',
    },
  },
  cli: {
    log: jest.fn(),
  },
};

beforeAll(() => {
  if (!fs.existsSync(slsDir)) {
    fs.mkdirSync(slsDir);
  }
});

describe('serverless-plugin-write-env-vars', () => {
  describe('constructor', () => {
    it('sets the correct defaults', () => {
      const plugin = new Plugin(slsDefaults, {});

      expect(plugin.options.stage).toBe('foo');
      expect(plugin.options.region).toBe('eu-west-1');
      expect(plugin.newTemplate).toBe(`${templatePrefix}.json`);
      expect(plugin.oldTemplate).toBe(`${templatePrefix}.org.json`);
    });

    it('registers the appropriate hooks', () => {
      const plugin = new Plugin(slsDefaults, {});

      expect(typeof plugin.hooks['before:diff:diff']).toBe('function');
      expect(typeof plugin.hooks['diff:diff']).toBe('function');
    });
  });

  describe('downloadTemplate', () => {
    afterEach(() => AWS.restore('CloudFormation'));

    describe('with successful CloudFormation call', () => {
      beforeEach(() => AWS.mock('CloudFormation', 'getTemplate', {
        TemplateBody: '{"foo":"bar"}',
      }));

      it('downloads currently deployed template', () => {
        const plugin = new Plugin(slsDefaults, {});

        return plugin.downloadTemplate()
          .then(() => fs.readFile(`${templatePrefix}.org.json`, { encoding: 'utf8' })
            .then((data) => expect(data).toMatchSnapshot()));
      });

      it('downloads currently deployed template with custom stackName', () => {
        const sls = { ...slsDefaults };
        sls.service.provider = { stackName: 'foo-r' };
        const plugin = new Plugin(sls, {});

        return plugin.downloadTemplate()
          .then(() => fs.readFile(`${templatePrefix}.org.json`, { encoding: 'utf8' })
            .then((data) => expect(data).toMatchSnapshot()));
      });

      it('downloads currently deployed template with preV1Resources flag', () => {
        const sls = { ...slsDefaults };
        sls.service.provider = { preV1Resources: true };
        const plugin = new Plugin(sls, {});

        return plugin.downloadTemplate()
          .then(() => fs.readFile(`${templatePrefix}.org.json`, { encoding: 'utf8' })
            .then((data) => expect(data).toMatchSnapshot()));
      });
    });

    describe('with unsuccessful CloudFormation call', () => {
      beforeEach(() => AWS.mock('CloudFormation', 'getTemplate', (Object, callback) => callback(new Error('Stack with id foo-foo does not exist'), null)));

      it('could not download deployed template', () => {
        const plugin = new Plugin(slsDefaults, {});

        return plugin.downloadTemplate()
          .catch((err) => expect(err).toMatchSnapshot());
      });
    });
  });

  describe('diff', () => {
    describe('successfully triggers diff', () => {
      beforeEach(() => fs.writeFile(exampleTemplate, '{"foo":"foo"}')
        .then(() => fs.writeFile(exampleOrgTemplate, '{"foo":"bar"}')));

      it('runs diff with defaults', () => {
        const plugin = new Plugin(slsDefaults, {});

        return plugin.diff()
          .then((data) => expect(data).toMatchSnapshot());
      });
    });

    describe('runs diff without changes', () => {
      beforeEach(() => fs.writeFile(exampleTemplate, '{"foo":"foo"}')
        .then(() => fs.writeFile(exampleOrgTemplate, '{"foo":"foo"}')));

      it('successfully triggers diff', () => {
        const plugin = new Plugin(slsDefaults, {});

        return plugin.diff()
          .then((data) => expect(data).toMatchSnapshot());
      });
    });

    describe('could not find locally compiled template', () => {
      beforeEach(() => fs.unlink(exampleTemplate));

      it('unsuccessfully triggers diff', () => {
        const plugin = new Plugin(slsDefaults, {});

        return plugin.diff()
          .catch((err) => expect(err).toMatchSnapshot());
      });
    });
  });
});
