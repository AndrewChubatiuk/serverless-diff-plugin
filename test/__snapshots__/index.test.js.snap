exports[`serverless-plugin-write-env-vars diff could not find locally compiled template unsuccessfully triggers diff 1`] = `".serverless/cloudformation-template-update-stack.json could not be found: run \"sls deploy --noDeploy\" first."`;

exports[`serverless-plugin-write-env-vars diff runs diff with custom localTemplate template successfully triggers diff 1`] = `
" {
[31m-  foo: \"custom\"[39m
[32m+  foo: \"foo\"[39m
 }
"
`;

exports[`serverless-plugin-write-env-vars diff runs diff without changes successfully triggers diff 1`] = `
" undefined
"
`;

exports[`serverless-plugin-write-env-vars diff successfully triggers diff runs diff with custom diff tool 1`] = `
"1c1
< {\"foo\":\"bar\"}
\\ No newline at end of file
---
> {\"foo\":\"foo\"}
\\ No newline at end of file
"
`;

exports[`serverless-plugin-write-env-vars diff successfully triggers diff runs diff with defaults 1`] = `
" {
[31m-  foo: \"bar\"[39m
[32m+  foo: \"foo\"[39m
 }
"
`;

exports[`serverless-plugin-write-env-vars downloadTemplate with successful CloudFormation call downloads currently deployed template 1`] = `
"{
  \"foo\": \"bar\"
}"
`;

exports[`serverless-plugin-write-env-vars downloadTemplate with successful CloudFormation call downloads currently deployed template with custom stackName 1`] = `
"{
  \"foo\": \"bar\"
}"
`;

exports[`serverless-plugin-write-env-vars downloadTemplate with unsuccessful CloudFormation call could not download deployed template 1`] = `"Stack with id foo-foo does not exist"`;
